### PR TITLE
feat: generate scholarship certificate

### PR DESCRIPTION
**Pull request by RickaPrincy**

### CODE REVIEW

File: src/operations/utils/GenCertificateButton.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import {useNotify} from "ra-core";
+import {studenstFileApi} from "../../providers/api";
+import {Download} from "@mui/icons-material";
+import {Button} from "ra-ui-materialui";
+import {useCallback} from "react";
+
+export function GenCertificateButton({studentId}) {
+  const notify = useNotify();
+  const FILE_NAME = "Certificat_Scolarité.pdf";
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: get constants out of the component

File: src/operations/utils/GenCertificateButton.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      .getStudentScholarshipCertificate(studentId, {
+        responseType: "arraybuffer",
+      })
+      .then((response) => {
+        const link = document.createElement("a");
+        link.href = window.URL.createObjectURL(
+          new Blob([response.data], {type: "application/pdf"})
+        );
+        link.download = FILE_NAME;
+        document.body.appendChild(link);
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: /!\ AVOID MUTATING THE DOM DIRECTLY, use React ref instead

File: src/operations/utils/GenCertificateButton.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import {studenstFileApi} from "../../providers/api";
+import {Download} from "@mui/icons-material";
+import {Button} from "ra-ui-materialui";
+import {useCallback} from "react";
+
+export function GenCertificateButton({studentId}) {
+  const notify = useNotify();
+  const FILE_NAME = "Certificat_Scolarité.pdf";
+
+  const generate = useCallback(() => {
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: only memoize performant critical function.
here `generate` fn fetches data from the api and appends an anchor to the body which doesn't evolve any expensive compution at all 

/!\ `generate` is not explicit at all and doesn't actually generate anything,
It fetches something from the api instead, `get...` would be more appropriate

File: src/operations/utils/GenCertificateButton.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    notify("Certificat de scolarité en cours de téléchargement", {
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: notify fn should be part of the [deps,]

File: src/operations/utils/GenCertificateButton.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import {useNotify} from "ra-core";
+import {studenstFileApi} from "../../providers/api";
+import {Download} from "@mui/icons-material";
+import {Button} from "ra-ui-materialui";
+import {useCallback} from "react";
+
+export function GenCertificateButton({studentId}) {
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: `GenCertificateButton` is proper to student and is too dependent to student to be a utility cmp.
move it inside `students` module under `components` instead

the same goes for `SelectSpecialization`

File: src/operations/students/index.js [hunk 6]
```diff
@@ -6,7 +6,7 @@ import StudentCreate from "./StudentCreate";
 
 const students = {
   list: StudentList,
-  edit: <ProfileEdit isOwnProfile={false} />,
+  edit: <ProfileEdit isOwnProfile={false} isStudent={true} />,
```
Comment (YumeT023) [inline src/operations/students/index.js]: _STYLE_

...`={true}` can be omitted

File: src/operations/utils/SelectSpecialization.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    case SpecializationField.EL:
+      return "Écosystéme Logiciel (EL)";
+    case SpecializationField.TN:
+      return "Transformation Numérique (TN)";
+    default:
+      return "Tronc commun";
+  }
+}
+
+export function SelectSpecialization({ignoreRole = false, ...props}) {
```
Comment (YumeT023) [inline src/operations/utils/SelectSpecialization.js]: conditionally render `SelectSpecialization` instead of passing `ignoreRole`
add some comments when using rest parameters `...props` as the component doesn't tell exactly what `props` would be, you could pass it through `selectFieldProps` maybe ...

File: src/operations/utils/SelectSpecialization.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import {SpecializationField, WhoamiRoleEnum} from "@haapi/typescript-client";
+import {SelectInput} from "react-admin";
+import authProvider from "../../providers/authProvider";
+
+export function getSpecializationValue(specialization) {
```
Comment (YumeT023) [inline src/operations/utils/SelectSpecialization.js]: is `getSpecializationValue` used anywhere else ?

File: src/operations/utils/SelectSpecialization.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      return "Transformation Numérique (TN)";
+    default:
+      return "Tronc commun";
+  }
+}
+
+export function SelectSpecialization({ignoreRole = false, ...props}) {
+  const isStudent = authProvider.getCachedRole() == WhoamiRoleEnum.STUDENT;
+
+  const SpecializationList = Object.entries(SpecializationField).map((el) => ({
```
Comment (YumeT023) [inline src/operations/utils/SelectSpecialization.js]: /!\ case is incorrect
map the `SpecializationField` to a constant and out of this component instead as it is not dependent to the cmp

el[0] el[1] is semantically incorrect

name them in the (anonym) fn of `map` using object destructuring

File: src/operations/utils/SelectSpecialization.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    value: el[0],
+    name: getSpecializationValue(el[1]),
+  }));
+
+  return ignoreRole || isStudent ? (
```
Comment (YumeT023) [inline src/operations/utils/SelectSpecialization.js]: render `SelectInput` directly, conditionally render the `cmp` instead

File: src/operations/utils/GenCertificateButton.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      autoHideDuration: 5000,
+    });
+    studenstFileApi()
+      .getStudentScholarshipCertificate(studentId, {
+        responseType: "arraybuffer",
+      })
+      .then((response) => {
+        const link = document.createElement("a");
+        link.href = window.URL.createObjectURL(
+          new Blob([response.data], {type: "application/pdf"})
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: correctly check if response is not null and that data (size/len/byteLen is not 0)
to avoid any potential `NPE` or any other Runtime err

File: src/operations/utils/GenCertificateButton.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        );
+        link.download = FILE_NAME;
+        document.body.appendChild(link);
+        link.click();
+        document.body.removeChild(link);
+      })
+      .catch(() =>
+        notify("Échec de téléchargement. Veuillez réessayer", {
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: notify is always called with
```js
{autoHideDuration: 5000}
```
option

/!\ avoid repetition
instead, create a hook wrapper (another hook that wraps the notify() that only take the msg arg)

File: src/operations/utils/GenCertificateButton.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+          autoHideDuration: 5000,
+        })
+      );
+  }, [studentId]);
+
+  return (
+    <Button label="Certificat" onClick={generate}>
+      <Download />
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: alias `Download` cmp to be more explicit
`DownloadIcon` is better

File: src/operations/profile/ProfileShow.js [hunk 14]
```diff
@@ -53,6 +54,9 @@ const renderSex = ({sex}) => {
   }
 };
 
+const renderSpecialization = ({specialization_field}) =>
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: `renderSpecialization` should take directly the `specialization_field` in param instead of the `record` itself

the same goes for most of the `renderers`
/!\ put a TODO on 'em

File: src/operations/profile/ProfileShow.js [hunk 15]
```diff
@@ -259,7 +269,8 @@ export const ProfileLayout = () => {
 };
 
 const ProfileShow = () => {
-  const id = authProvider.getCachedWhoami().id;
+  const {role, id} = authProvider.getCachedWhoami();
+  const isStudent = role === WhoamiRoleEnum.STUDENT;
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: /!\ avoid repetition
extract these role checks to a custom react hook
as It is everywhere in the project
put a refactor TODO for now

File: src/operations/profile/ProfileShow.js [hunk 16]
```diff
@@ -147,7 +151,7 @@ const Title = ({children}) => (
   </Typography>
 );
 
-export const ProfileLayout = () => {
+export const ProfileLayout = ({isStudent = false}) => {
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js:155]: either don't export `ProfileLayout` or move it in `components` subdir
If used accross operations, move it in operations/common/components

/!\ put a refactor TODO

File: src/operations/utils/GenCertificateButton.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
+import {useNotify} from "ra-core";
+import {studenstFileApi} from "../../providers/api";
```
Comment (YumeT023) [inline src/operations/utils/GenCertificateButton.js]: Remove unused imports

**Review by YumeT023 — CHANGES_REQUESTED** [review]

File: src/operations/utils/SelectSpecialization.js (hunk 8 above)
Comment (RickaPrincy) [reply src/operations/utils/SelectSpecialization.js]: yes ... used in profilShow

File: src/hooks/useNotifyMessage.js [hunk 18]
```diff
@@ -0,0 +1,6 @@
+import {useNotify} from "react-admin";
+
+export const useNotifyMessage = (config = {}) => {
```
Comment (YumeT023) [inline src/hooks/useNotifyMessage.js]: instead, alias the `useNotify` of react-admin as `_useNotify` or `useNotifyInternal`
then rename `useNotifyMessage` into `useNotify` as usual (rename the file too)

File: src/hooks/useRole.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import {WhoamiRoleEnum} from "@haapi/typescript-client";
+import authProvider from "../providers/authProvider";
+
+export function useRole() {
+  const auth = authProvider.getCachedWhoami();
+  return {
+    isStudent: auth.role === WhoamiRoleEnum.STUDENT,
+    isManager: auth.role === WhoamiRoleEnum.MANAGER,
+    isTeacher: auth.role === WhoamiRoleEnum.TEACHER,
+    ...auth,
```
Comment (YumeT023) [inline src/hooks/useRole.js]: `useRole` shouldn't expose any other thing apart from the `role`

you could return the `is...` and the `role` itself but not the `auth`

File: src/operations/profile/ProfileEdit.js [hunk 20]
```diff
@@ -2,48 +2,57 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
+      <SimpleForm toolbar={<EditToolBar />}>
+        <TextInput
+          source="ref"
+          label="Référence"
+          fullWidth
+          disabled={isOwnProfile}
+        />
+        <TextInput source="first_name" label="Prénom·s" fullWidth />
+        <TextInput source="last_name" label="Nom·s" fullWidth />
+        {(isStudent || role.isStudent) && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: what does this mean ? `isStudent` then role.isStudent ? ... arrange this a bit

File: src/hooks/useRole.js [hunk 21]
```diff
@@ -0,0 +1,12 @@
+import {WhoamiRoleEnum} from "@haapi/typescript-client";
+import authProvider from "../providers/authProvider";
+
+export function useRole() {
+  const auth = authProvider.getCachedWhoami();
+  return {
```
Comment (YumeT023) [inline src/hooks/useRole.js]: react will evaluate the 
```js
auth.role === _
```
directly, ... instead, turn isStudent, isManager, isTeacher into (anonymous) fn

File: src/hooks/useRole.js [hunk 22]
```diff
@@ -0,0 +1,12 @@
+import {WhoamiRoleEnum} from "@haapi/typescript-client";
+import authProvider from "../providers/authProvider";
+
+export function useRole() {
```
Comment (YumeT023) [inline src/hooks/useRole.js]: /!\ move  `useRole` hooks to 
security/hooks/useRole as it's part of the `core` and related to security/permission

File: src/operations/profile/ProfileEdit.js [hunk 23]
```diff
@@ -2,48 +2,57 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
+          <SelectSpecialization disabled={isOwnProfile} />
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: `ProfileEdit` ... is it still necessary to check if it is my own profile or not ?
You shouldn't even get to `ProfileEdit` unless its your own profile
/!\ re-adjust this or add some comments to explain it

File: src/operations/profile/ProfileEdit.js [hunk 24]
```diff
@@ -2,48 +2,57 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
+  const role = useRole();
+
+  return (
+    <CustomEdit title="Modifier le profil" transform={usertoUserApi}>
+      <SimpleForm toolbar={<EditToolBar />}>
+        <TextInput
+          source="ref"
+          label="Référence"
+          fullWidth
+          disabled={isOwnProfile}
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: ... ProfileEdit should only be accessed by its owner, would have been okay if the cmp was `ProfileShow`

File: src/operations/profile/ProfileEdit.js [hunk 25]
```diff
@@ -2,48 +2,57 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
-      <SexRadioButton />
-      <TextInput source="phone" label="Téléphone" fullWidth />
-      <TextInput
-        source="nic"
-        label="Numéro CIN"
-        fullWidth
-        validate={maxLength(
-          12,
-          "Le numéro CIN ne doit pas dépasser 12 caractères."
+const ProfileEdit = ({isOwnProfile, isStudent}) => {
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js:14]: ... `isOwnProfile` in `ProfileEdit` doesn't make any sense

File: src/operations/profile/ProfileShow.js [hunk 26]
```diff
@@ -259,7 +280,7 @@ export const ProfileLayout = () => {
 };
 
 const ProfileShow = () => {
-  const id = authProvider.getCachedWhoami().id;
+  const {isStudent, id} = useRole();
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: /!\  `useRole` must _ONLY_ check and expose current user `role`

File: src/operations/students/StudentCreate.js [hunk 27]
```diff
@@ -53,6 +54,7 @@ const StudentCreate = () => {
         <TextInput source="last_name" label="Nom" fullWidth required />
         <SexRadioButton />
         <TextInput source="phone" label="Téléphone" fullWidth />
+        <SelectSpecialization ignoreRole={true} />
```
Comment (YumeT023) [inline src/operations/students/StudentCreate.js:57]: what is `ignoreRole` prop for ?

File: src/operations/students/StudentShow.js [hunk 28]
```diff
@@ -39,7 +42,7 @@ const StudentShow = () => {
       title="Étudiants"
       actions={role === WhoamiRoleEnum.MANAGER && <ActionsOnShow />}
     >
-      <ProfileLayout />
+      <ProfileLayout isStudent={true} />
```
Comment (YumeT023) [inline src/operations/students/StudentShow.js]: /!\ use implicit truthy bool prop
`<Cmp isStudent />`

File: src/operations/students/StudentShow.js [hunk 29]
```diff
@@ -19,14 +19,17 @@ const ActionsOnShow = ({basePath, data, resource}) => {
...
-          to={`/students/${record.id}/fees`}
-        >
-          <AttachMoney />
-        </Button>
+        <>
+          <Button
+            label="Frais"
+            aria-label="fees"
+            component={Link}
+            to={`/students/${record.id}/fees`}
```
Comment (YumeT023) [inline src/operations/students/StudentShow.js]: rename `record` => `student`

File: src/operations/students/StudentShow.js [hunk 30]
```diff
@@ -39,7 +42,7 @@ const StudentShow = () => {
       title="Étudiants"
       actions={role === WhoamiRoleEnum.MANAGER && <ActionsOnShow />}
```
Comment (YumeT023) [inline src/operations/students/StudentShow.js]: use the `useRole` hook

File: src/operations/students/components/GenCertificateButton.js [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+    notify("Certificat de scolarité en cours de téléchargement", {
+      autoHideDuration: 5000,
+    });
+    studenstFileApi()
+      .getStudentScholarshipCertificate(studentId, {
+        responseType: "arraybuffer",
+      })
+      .then((response) => {
+        if (!response.data) {
+          throw new Error("The response data from the server is null");
```
Comment (YumeT023) [inline src/operations/students/components/GenCertificateButton.js]: response.data may be defined but is 0 byte.

improve the checks.
- data is truthy
- data `size/byteLength/length` (rtfd which prop is associated the file object from server `ArrayBuffer`)
   - should be `> 0` 

those things may lead to unexpected  behaviour if not checked properly

File: src/operations/students/components/GenCertificateButton.js [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+        }
+        certificateLink.current.href = window.URL.createObjectURL(
```
Comment (YumeT023) [inline src/operations/students/components/GenCertificateButton.js]: `current` prop of `certificateLink` is accessed multiple times,
instead, put it in a variable `certificateLink`, suffix react ref with `ref`, `certificateLinkRef` here

File: src/operations/students/components/GenCertificateButton.js (hunk 31 above)
Comment (YumeT023) [inline src/operations/students/components/GenCertificateButton.js]: Do not throw an err when the data is not gud, instead report it with a `toast` notif or smth
as network conn is not as reliable as u'd expect...

File: src/operations/students/components/GenCertificateButton.js [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+          new Blob([response.data], {type: "application/pdf"})
+        );
+        certificateLink.current.download = FILE_NAME;
+        certificateLink.current.click();
+      })
+      .catch(() =>
```
Comment (YumeT023) [inline src/operations/students/components/GenCertificateButton.js]: always add braces when you're not returning anything from the anonymous_fn

File: src/operations/students/components/GenCertificateButton.js [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+        notify("Échec de téléchargement. Veuillez réessayer", {
+          autoHideDuration: 5000,
+        })
+      );
+  };
+
+  return (
+    <>
```
Comment (YumeT023) [inline src/operations/students/components/GenCertificateButton.js]: but it inside a kind of container, `div` or something, ...

File: src/operations/students/components/SelectSpecialization.js [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+import {SelectInput} from "react-admin";
+
+const SPECIALIZATION_LIST = Object.entries(SpecializationField).map(
+  ([value, name]) => ({
+    value,
+    name: getSpecializationValue(name),
+  })
+);
+
+export function getSpecializationValue(specialization) {
```
Comment (YumeT023) [inline src/operations/students/components/SelectSpecialization.js]: instead, make a _STATIC_ constant object for mapping
then `SPECIALIZATIONS[key] ?? "Tronc commun"`

Comment (YumeT023) [inline src/operations/students/components/SelectSpecialization.js]: why is this `fn` exported outside of `SelectSpecialization` scope ?

File: src/operations/students/components/index.js [hunk 36]
```diff
@@ -0,0 +1,2 @@
+export * from "./GenCertificateButton";
```
Comment (YumeT023) [inline src/operations/students/components/index.js]: rename this `cmp`, as it is not really a button, simply `GetCertificate` (also, it is not about generating anything)

File: src/operations/students/index.js [hunk 37]
```diff
@@ -6,7 +6,7 @@ import StudentCreate from "./StudentCreate";
 
 const students = {
   list: StudentList,
-  edit: <ProfileEdit isOwnProfile={false} />,
+  edit: <ProfileEdit isOwnProfile={false} isStudent />,
```
Comment (YumeT023) [inline src/operations/students/index.js:10]: You may check role inside of `<ProfileEdit /> directly

what is `isOwnProfile` for ?

File: src/operations/profile/ProfileShow.js [hunk 38]
```diff
@@ -197,12 +204,23 @@ export const ProfileLayout = () => {
                 color={PALETTE_COLORS.yellow}
               />
               <TextField source="role" label="Rôle" />
+              {(isStudent || role.isStudent) && (
+                <FunctionField
+                  label="Parcours de Spécialisation"
+                  render={(record) =>
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: /!\ put a TODO to rename these `record` => the associated data like (`student`, `fee`, ...etc)

File: src/hooks/index.js [hunk 39]
```diff
@@ -1,3 +1,5 @@
 export * from "./useStudentRef";
 export * from "./useCreateFees";
 export * from "./useToggle";
+export * from "./useRole";
```
Comment (YumeT023) [inline src/hooks/index.js]: useRole hook is part of the `core`, not related to toggle or notification hook, move it in security/hooks instead

/!\ put TODO to move the `useCreateFees` and `useStudentRef` too

**Review by YumeT023 — CHANGES_REQUESTED** [review]

File: src/operations/students/components/SelectSpecialization.js (hunk 35 above)
Comment (RickaPrincy) [reply src/operations/students/components/SelectSpecialization.js]: Use to show the value of specialization on profileShow

File: src/operations/students/index.js (hunk 37 above)
Comment (RickaPrincy) [reply src/operations/students/index.js:10]: `isStudent` is used to say that the profile we wanna edit is a student's profile  (we need it 'cause Manager can edit student's profile, teacher's profile...) 
Samething for  isOwnProfile ...

File: src/operations/profile/ProfileShow.js (hunk 38 above)
Comment (RickaPrincy) [reply src/operations/profile/ProfileShow.js]: But ProfilShow is for all resources ... we cannot determine if it's student or fee ....

File: src/operations/profile/ProfileEdit.js (hunk 25 above)
Comment (RickaPrincy) [reply src/operations/profile/ProfileEdit.js:14]: like .... I can edit your profile ...

File: src/operations/profile/ProfileShow.js (hunk 38 above)
Comment (YumeT023) [reply src/operations/profile/ProfileShow.js]: I mean, Profile means `user` ... not fee in any case
And here, the fact is that you are only rendering the fn field when the user is a student

student/teacher/manager record => user
fee record => fee
`...`

File: src/operations/students/components/SelectSpecialization.js (hunk 35 above)
Comment (YumeT023) [reply src/operations/students/components/SelectSpecialization.js]: why isn't it named like the other ? 
why isn't it named `renderSpec...` ?

File: src/operations/students/index.js (hunk 37 above)
Comment (YumeT023) [reply src/operations/students/index.js:10]: Instead make another variant of ProfileEdit for Teacher and Student instead of having lots of flag here and there

File: src/operations/profile/ProfileShow.js (hunk 38 above)
Comment (RickaPrincy) [reply src/operations/profile/ProfileShow.js]: ahh ... ok ok

Comment (RickaPrincy) [reply src/operations/profile/ProfileShow.js]: user seems logic ...

File: src/operations/students/index.js (hunk 37 above)
Comment (RickaPrincy) [reply src/operations/students/index.js:10]: Why not TODO ? the PR will be very huge for that 
Like I put that one another PR ...

File: src/operations/students/components/SelectSpecialization.js (hunk 35 above)
Comment (RickaPrincy) [reply src/operations/students/components/SelectSpecialization.js]: getSpecializationValue is no longer present as said on the other change request, now it's a simple `const` as you said

File: src/__tests__/GetCertificate.cy.js [hunk 40]
```diff
@@ -0,0 +1,39 @@
+import {mount, unmount} from "@cypress/react";
+import App from "../App";
+import {student1} from "./credentials";
+import specTitle from "cypress-sonarqube-reporter/specTitle";
+import {student1Mock, whoamiStudentMock} from "./mocks/responses";
+
+const FILE_NAME = "Cetificat_scolarité.pdf";
+
+describe(
+  specTitle("Student and Manager Can get a certificate for students"),
```
Comment (YumeT023) [inline src/__tests__/GetCertificate.cy.js]: More concise description: Student certificate

Comment (YumeT023) [inline src/__tests__/GetCertificate.cy.js]: test both `student` and `manager` case as you said in the spec_title

File: src/__tests__/GetCertificate.cy.js [hunk 41]
```diff
@@ -0,0 +1,39 @@
...
+      cy.get("#password").type(student1.password);
+      cy.get("button").contains("Connexion").click();
+      cy.intercept("GET", `/whoami`, whoamiStudentMock).as("getWhoami");
+      cy.intercept("GET", `/students/${student1Mock.id}`, student1Mock).as(
+        "getStudent1"
+      );
+      cy.wait("@getWhoami");
+      cy.get("body").click();
+    });
+    it("can edit students", () => {
```
Comment (YumeT023) [inline src/__tests__/GetCertificate.cy.js]: Edit student ?

File: src/__tests__/GetCertificate.cy.js [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+        "GET",
+        `/students/${student1Mock.id}/scholarship_certificate/raw`,
+        {fixture: "students/certificate.pdf"}
+      ).as("downloadCertificate");
+      cy.get('[data-testid="get-certificate-btn"]').click();
+      cy.wait("@downloadCertificate");
+      cy.get('[data-testid="certificate-link"]')
+        .should("not.be.visible")
+        .and("have.attr", "href")
+        .and("include", "blob");
```
Comment (YumeT023) [inline src/__tests__/GetCertificate.cy.js]: construct a blob from the href uri and test that its size is > 0
also test when the retrieved file is null (the server responded null) or when the returned file is == 0

File: src/operations/profile/ProfileEdit.js [hunk 43]
```diff
@@ -2,48 +2,52 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
+  const role = useRole();
+
+  return (
+    <CustomEdit title="Modifier le profil" transform={usertoUserApi}>
+      <SimpleForm toolbar={<EditToolBar />}>
+        <TextInput source="ref" label="Référence" fullWidth disabled={true} />
+        <TextInput source="first_name" label="Prénom·s" fullWidth />
+        <TextInput source="last_name" label="Nom·s" fullWidth />
+        {(isStudent || role.isStudent()) && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: what is isStudent || role.isStudent() ? ... get rid of such ambiguity

File: src/operations/profile/ProfileShow.js [hunk 44]
```diff
@@ -197,12 +203,23 @@ export const ProfileLayout = () => {
                 color={PALETTE_COLORS.yellow}
               />
               <TextField source="role" label="Rôle" />
+              {(isStudent || role.isStudent()) && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: ambiguous

File: src/operations/profile/ProfileShow.js [hunk 45]
```diff
@@ -272,6 +293,7 @@ const ProfileShow = () => {
             to={`/profile/${id}/edit`}
             data-testid="profile-edit-button"
           />
+          {isStudent() && <GetCertificate studentId={id} />}
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: the `role` context is needed here
`role.isStudent()` is more explicit and helps avoid any ambiguity such as:
- The shown profile is a student
- The user who views the profile is a student

File: src/operations/students/StudentShow.js [hunk 46]
```diff
@@ -9,37 +9,36 @@ import {
...
+          <GetCertificate studentId={student.id} />
+        </>
       )}
     </TopToolbar>
   );
 };
 
 const StudentShow = () => {
-  const role = authProvider.getCachedRole();
+  const {isManager} = useRole();
```
Comment (YumeT023) [inline src/operations/students/StudentShow.js]: always include the _role_ ctx , role.isManager()

File: src/operations/students/components/SelectSpecialization.js [hunk 47]
```diff
@@ -0,0 +1,25 @@
+import {SelectInput} from "react-admin";
+
+export const SPECIALIZATION_VALUE = {
+  EL: "Écosystéme Logiciel (EL)",
+  TN: "Transformation Numérique (TN)",
+  COMMON_CORE: "Tronc commun",
+};
+
+export function SelectSpecialization(props) {
+  const specializationChoices = Object.entries(SPECIALIZATION_VALUE).map(
```
Comment (YumeT023) [inline src/operations/students/components/SelectSpecialization.js]: again, specializationChoices can be computed outside of this component

File: src/operations/students/index.js (hunk 37 above)
Comment (YumeT023) [inline src/operations/students/index.js:10]: put a TODO
/!\ Create ProfileEdit variant: TeacherProfileEdit, StudentProfileEdit instead of having a flag that says if the ProfileEdit component should show the student version ...

**Review by YumeT023 — CHANGES_REQUESTED** [review]

**Comment by YumeT023** [comment]
test commit is scoped under test
e.g: test: get certificate

File: src/operations/profile/ProfileEdit.js (hunk 43 above)
Comment (RickaPrincy) [reply src/operations/profile/ProfileEdit.js]: I think I'll put some comments then... anyway, isStudent is passed through the cmp to say that the record is a student ( when manager want to watch a profile os student, or teachers)
role.isStudent() is to check if the current logged user is a student...

File: src/__tests__/GetCertificate.cy.js (hunk 41 above)
Comment (RickaPrincy) [reply src/__tests__/GetCertificate.cy.js]: my bad ... copy paste error

File: src/__tests__/GetCertificate.cy.js [hunk 48]
```diff
@@ -0,0 +1,99 @@
+import {mount, unmount} from "@cypress/react";
```
Comment (YumeT023) [inline src/__tests__/GetCertificate.cy.js:1]: LGTM!

Separate interaction from assertion (separate click, thing... from contains, and eq)

File: src/operations/profile/ProfileEdit.js [hunk 49]
```diff
@@ -2,48 +2,55 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
+
+  return (
+    <CustomEdit title="Modifier le profil" transform={usertoUserApi}>
+      <SimpleForm toolbar={<EditToolBar />}>
+        <TextInput source="ref" label="Référence" fullWidth disabled={true} />
+        <TextInput source="first_name" label="Prénom·s" fullWidth />
+        <TextInput source="last_name" label="Nom·s" fullWidth />
+        {
+          // Shown if the record of user is a student or if the current logged user is a student
+          (isStudent || role.isStudent()) && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: Remove the useless comment
put this in a variable before render

File: src/operations/profile/ProfileShow.js [hunk 50]
```diff
@@ -197,12 +203,26 @@ export const ProfileLayout = () => {
                 color={PALETTE_COLORS.yellow}
               />
               <TextField source="role" label="Rôle" />
+              {
+                // Shown if the record of user is a student or if the current logged user is a student
+                (isStudent || role.isStudent()) && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: Remove the useless comment
put the condition in a well-named variable

**Review by YumeT023 — CHANGES_REQUESTED** [review]

File: src/operations/profile/ProfileEdit.js [hunk 51]
```diff
@@ -2,48 +2,51 @@ import {DateInput, maxLength, SimpleForm, TextInput} from "react-admin";
...
-        fullWidth
-        disabled={isOwnProfile}
-      />
-      <StatusRadioButton disabled={isOwnProfile} />
-    </SimpleForm>
-  </CustomEdit>
-);
+const ProfileEdit = ({isOwnProfile, isStudent}) => {
+  const role = useRole();
+  const isRecordOfStudent = isStudent || role.isStudent();
```
Comment (YumeT023) [inline src/operations/profile/ProfileEdit.js]: not clear enough

File: src/operations/profile/ProfileShow.js [hunk 52]
```diff
@@ -197,12 +204,23 @@ export const ProfileLayout = () => {
                 color={PALETTE_COLORS.yellow}
               />
               <TextField source="role" label="Rôle" />
+              {isRecordOfStudent && (
```
Comment (YumeT023) [inline src/operations/profile/ProfileShow.js]: not clear enough

**Review by YumeT023 — CHANGES_REQUESTED** [review]

**Review by YumeT023 — APPROVED** [review]
